larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,355 @@
+/**
+ *  @file   larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc
+ *
+ *  @brief  Implementation file for the reclustering algorithm class.
+ *          Tries multiple 3D clustering algorithms and reclusters Pfos is a good one is found.
+ *          Reclustered Pfos will: have no vertex, not be characterised as track/shower, not be part of any particle hierarchy.
+ *          These properties will need to be added by succeeding algorithms.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+
+#include "larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+ThreeDMultiReclusteringAlgorithm::ThreeDMultiReclusteringAlgorithm() :
+    m_clusterListNames {{TPC_3D, {}}, {TPC_VIEW_U, "ClustersU"}, {TPC_VIEW_V, "ClustersV"}, {TPC_VIEW_W, "ClustersW"}},
+    m_pFomAlgTool {nullptr}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDMultiReclusteringAlgorithm::Run()
+{
+    const PfoList *pPfos {nullptr};
+    PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=,
+                                    PandoraContentApi::GetList(*this, m_pfoListName, pPfos));
+    if (!pPfos || pPfos->empty())
+        return STATUS_CODE_SUCCESS;
+
+    // Ask the FOM alg tool which pfos to recluster
+    PfoList pfosToRecluster;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->GetPfosToRecluster(pPfos, pfosToRecluster));
```

Comment:
Do we only ever want to calculate a single FOM?

Reply:
I was more thinking that we might want to run multiple tools, each of which calculates a single FOM.  Is that not ever going to be the case?